### PR TITLE
Unified eq and compare

### PR DESCRIPTION
**Pull request by isuruf**
@htfy96, thanks for the idea. This makes the code simpler and avoids duplication. Can you review?

### CODE REVIEW

**Comment by htfy96** [comment]
Great! I'll look into the code tomorrow.

File: symengine/expression.h [hunk 1]
```diff
@@ -179,6 +179,22 @@ inline Expression expand(const Expression &arg)
...
+}
+
+template <typename T, typename U,
+          typename = enable_if_t<std::is_same<U, Expression>::value
+                                 and std::is_same<T, Expression>::value>>
+inline int unified_compare(const T &a, const U &b)
+{
+    return unified_compare(a.get_basic(), b.get_basic());
+}
+
```
Comment (htfy96) [inline symengine/expression.h]: Would
```
inline bool unified_eq(const Expression &a, const Expression &b)
```
be a better choice?
Pro: 
 - Reduce compile time

Possible behavior change:
 - Overloading precedence: non-template function > template function. Changing it to non-template version might cause unexpected behavior when you want to match a subclass of `Expression` with template. In that case this version will have higher priority.

File: symengine/polynomial_multivariate.cpp [hunk 2]
```diff
@@ -166,35 +162,34 @@ vec_basic MultivariatePolynomial::get_args() const
...
     const MultivariatePolynomial &s
         = static_cast<const MultivariatePolynomial &>(o);
 
-    int cmp = set_compare<set_basic>(vars_, s.vars_);
+    if (vars_.size() != s.vars_.size())
+        return vars_.size() < s.vars_.size() ? -1 : 1;
+    if (dict_.size() != s.dict_.size())
+        return dict_.size() < s.dict_.size() ? -1 : 1;
+
+    int cmp = unified_compare(vars_, s.vars_);
```
Comment (htfy96) [inline symengine/polynomial_multivariate.cpp]: I think such comparison should be handled in a better way, e.g. 
```
return unified_compare(std::forward_as_tuple(vars_.size(), dict_.size(), vars_, dict_),
                std::forward_as_tuple(s.vars_.size(), s.dict_.size(), s.vars_, s.dict_));
```

A demo has just been implemented: https://github.com/htfy96/cpp-cmp/

**Comment by isuruf** [comment]
I am going to merge this and improve compare later.

**Pull request merged**